cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -723,53 +723,81 @@ func fixNodeGroupSize(context *context.AutoscalingContext, clusterStateRegistry
 	for _, unregisteredNode := range unregisteredNodes {
-		nodeGroup, err := context.CloudProvider.NodeGroupForNode(unregisteredNode.Node)
+		nodeGroup, err := a.CloudProvider.NodeGroupForNode(unregisteredNode.Node)
 		if err != nil {
 			klog.Warningf("Failed to get node group for %s: %v", unregisteredNode.Node.Name, err)
-			return removedAny, err
-		}
-		if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
-			klog.Warningf("No node group for node %s, skipping", unregisteredNode.Node.Name)
 			continue
 		}
 
 		maxNodeProvisionTime, err := csr.MaxNodeProvisionTime(nodeGroup)
 		if err != nil {
-			return removedAny, fmt.Errorf("failed to retrieve maxNodeProvisionTime for node %s in nodeGroup %s", unregisteredNode.Node.Name, nodeGroup.Id())
+			return false, fmt.Errorf("failed to retrieve maxNodeProvisionTime for node %s in nodeGroup %s", unregisteredNode.Node.Name, nodeGroup.Id())
 		}
 
 		if unregisteredNode.UnregisteredSince.Add(maxNodeProvisionTime).Before(currentTime) {
-			klog.V(0).Infof("Removing unregistered node %v", unregisteredNode.Node.Name)
-			size, err := nodeGroup.TargetSize()
-			if err != nil {
-				klog.Warningf("Failed to get node group size; unregisteredNode=%v; nodeGroup=%v; err=%v", unregisteredNode.Node.Name, nodeGroup.Id(), err)
-				continue
-			}
-			if nodeGroup.MinSize() >= size {
-				klog.Warningf("Failed to remove node %s: node group min size reached, skipping unregistered node removal", unregisteredNode.Node.Name)
-				continue
-			}
-			err = nodeGroup.DeleteNodes([]*apiv1.Node{unregisteredNode.Node})
-			csr.InvalidateNodeInstancesCacheEntry(nodeGroup)
-			if err != nil {
-				klog.Warningf("Failed to remove node %s: %v", unregisteredNode.Node.Name, err)
+			klog.V(0).Infof("Marking unregistered node %v for removal", unregisteredNode.Node.Name)
+			nodesToBeDeletedByNodeGroupId[nodeGroup.Id()] = append(nodesToBeDeletedByNodeGroupId[nodeGroup.Id()], unregisteredNode)
+		}
+	}
+
+	removedAny := false
+	for nodeGroupId, nodesToDelete := range nodesToBeDeletedByNodeGroupId {
+		nodeGroup := nodeGroups[nodeGroupId]
+
+		klog.V(0).Infof("Removing %v unregistered nodes for node group %v", len(nodesToDelete), nodeGroupId)
+		size, err := nodeGroup.TargetSize()
+		if err != nil {
+			klog.Warningf("Failed to get node group size; nodeGroup=%v; err=%v", nodeGroup.Id(), err)
+			continue
+		}
+		possibleToDelete := size - nodeGroup.MinSize()
+		if possibleToDelete <= 0 {
+			klog.Warningf("Node group %s min size reached, skipping removal of %v unregistered nodes", nodeGroupId, len(nodesToDelete))
+			continue
+		}
+		if len(nodesToDelete) > possibleToDelete {
+			klog.Warningf("Capping node group %s unregistered node removal to %d nodes, removing all %d would exceed min size constaint", nodeGroupId, possibleToDelete, len(nodesToDelete))
+			nodesToDelete = nodesToDelete[:possibleToDelete]
+		}
+		toDelete, err := a.fillInFailedNodesWithZeroOrMaxScaling(nodeGroup, toNodes(unregisteredNodes))
```

Comment:
nit: I think a bit more readable version would be something similar to the one below. I won't block the approval on this, but please consider what do you think about this:

```
// rename nodesToDelete loop variable to unregisteredNodes

nodesToDelete := toNodes(unregisteredNodes)

if isZeroOrMaxScalingEnabled(nodeGroup) {
  instances, err := nodeGroup.Nodes()
  // err handling
  nodesToDelete := instancesToFakeNodes(instances)
}

err = nodeGroup.DeleteNodes(toDelete)
```

Reply:
I think if we're going in this direction, I prefer to just inline fillInFailedNodesWithZeroOrMaxScaling. isZeroOrMaxScalingEnabled could return an error, so it's not super convenient to have it as a helper function (unfortunate that just retrieving options for a node group can error out, but there has to be a reason for that...)

Reply:
SGTM.